src/azul/collections.py

```diff
@@ -271,3 +273,102 @@ def to_dict(self):
             k: v.to_dict() if isinstance(v, NestedDict) else v
             for k, v in self.items()
         }
+
+
+class OrderedSet(MutableSet[K]):
+    """
+    A mutable set that maintains insertion order. Unlike similar implementations
+    of the same name that are floating around on the internet, it is not a
+    sequence.
+
+    >>> s = OrderedSet(['b', 'a', 'c', 'b'])
+    >>> s
+    OrderedSet(['b', 'a', 'c'])
+
+    >>> s.discard('a')
+    >>> s
+    OrderedSet(['b', 'c'])
+
+    >>> s.add('a')
+    >>> s
+    OrderedSet(['b', 'c', 'a'])
+    """
+
+    def __init__(self, members: Iterable[K] = (), /) -> None:
+        self.inner: dict[K, None] = dict.fromkeys(members)
+
+    def __repr__(self) -> str:
+        contents = repr(list(self)) if self else ''
+        return f'{type(self).__name__}({contents})'
+
+    def __iter__(self) -> Iterator[K]:
+        return iter(self.inner)
+
+    def __len__(self) -> int:
+        return len(self.inner)
+
+    def __eq__(self, other: Any) -> bool:
+        """
+        Symmetric property
+        >>> OrderedSet() == set()
+        True
+        >>> OrderedSet([1,2]) == set([1,2])
+        True
+
+        Transitive property
+        >>> s1 = OrderedSet([1])
+        >>> s2 = {1, 3}
+        >>> s3 = OrderedSet([1, 3, 4])
+        >>> s1 < s2
+        True
+        >>> s2 < s3
+        True
+        >>> s1 < s3
+        True
+
+        Commutative property
+        >>> s1 = OrderedSet(['3', '1', '3'])
+        >>> s2 = {'1', '3', '1'}
+        >>> s1 == s2
+        True
+        >>> s2 == s1
+        True
```

Comment:
Symmetric and commutative are [very nearly synonyms](https://math.stackexchange.com/q/803614/63189) so I don't think we should separate them into two cases. I believe "commutative" is more appropriate for methods that return sets (e.g. intersection) while "symmetric" should be used for methods like `__eq__` where the return value is not a set. Regardless, the current tests for equality are redundant between the two.

```suggestion
        Symmetric property
        >>> s1, s2 = OrderedSet(), set()
        >>> s1 == s2
        True
        >>> s2 == s1
        >>> s1, s2 = OrderedSet(['3', '1', '3']), {'1', '3', '1'}
        >>> s1 == s2
        True
        >>> s2 == s1
        True
```